src/routes.rs

```diff
@@ -35,10 +46,48 @@ pub async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert
     }
 }
 
+pub async fn create_account(
+    db_pool: deadpool_postgres::Pool,
+    auth_header: String,
+) -> Result<impl Reply, Rejection> {
+    let auth::Auth { username, password } = auth::validate_header(auth_header.as_str())?;
+    let pool = db_pool
+        .get()
+        .await
+        .map_err(|_| warp::reject::custom(InternalError))?;
+
+    let user = crate::db::get_user(&pool, &username)
+        .await
+        .map_err(|err| match err {
+            tokio_pg_mapper::Error::ColumnNotFound => warp::reject::not_found(),
+            _ => warp::reject::custom(InternalError),
+        });
+
+    if let Ok(_) = user {
+        return Err(warp::reject::custom(Conflict));
+    }
+
+    let salt: String = (0..10).map(|_| rand::random::<char>()).collect();
+
+    let mut hasher = Sha256::new();
+    let mut x = password.into_bytes();
+    x.extend(salt.trim().as_bytes());
+
+    hasher.update(x);
+    let result = &hasher.finalize()[..];
+
+    // todo(SirH): return the user (or at least the username)
+    let new_user = crate::db::create_user(&pool, &username, result, &salt)
+        .await
+        .map_err(|_| warp::reject::custom(InternalError))?;
+
+    Ok(crate::warp_reply!("Account created".to_owned(), CREATED))
+}
+
 pub async fn auth(
     db_pool: deadpool_postgres::Pool,
     auth_header: String,
-) -> std::result::Result<deadpool_postgres::Pool, Rejection> {
+) -> std::result::Result<impl Reply, Rejection> {
```

Comment:
This is incorrect and would prevent the intended use of `auth` which was so that .and_then(auth) could be prepended to any route which needed to only handle authenticated users
```rust
let routes = warp::any()
     .and_then(auth) // any .and_then s after this know for sure the user is authenticated
     .and_then(my_route)
     .recover(handler);
```
(rough code sample)

Reply:
Ngl, I didn't know that's what you were using auth for, thought it was more of just a simple handler for authenticating a user on a GET call

Reply:
thats what it is, it's just that you'd want to do stuff with the authenticated user instead of just checking theyre authenticated and then dropping the req